src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/ReadWriteSplittingPlugin.java

```diff
@@ -0,0 +1,271 @@
+   *            The connection where to set state.
+   * @param readOnly
+   *            The new read-only status.
+   */
+  void syncSessionState(JdbcConnection source, JdbcConnection target, boolean readOnly) {
+    try {
+      if (target != null) {
+        target.setReadOnly(readOnly);
+      }
+
+      if (source == null || target == null) {
+        return;
+      }
+
+      RuntimeProperty<Boolean> sourceUseLocalSessionState = source.getPropertySet().getBooleanProperty(PropertyKey.useLocalSessionState);
+      boolean prevUseLocalSessionState = sourceUseLocalSessionState.getValue();
+      sourceUseLocalSessionState.setValue(true);
+
+      target.setAutoCommit(source.getAutoCommit());
+      String db = source.getDatabase();
+      if (db != null && !db.isEmpty()) {
+        target.setDatabase(db);
+      }
+      target.setTransactionIsolation(source.getTransactionIsolation());
+      target.setSessionMaxRows(source.getSessionMaxRows());
+
+      sourceUseLocalSessionState.setValue(prevUseLocalSessionState);
+    } catch (SQLException e) {
+      // Do nothing. Reader connections must continue to "work" after swapping between writers and readers.
+    }
+  }
+
+  private synchronized void switchToReaderConnection(JdbcConnection currentConnection) throws SQLException {
+    if (this.readerConnection == null || this.readerConnection.isClosed()) {
+      initializeReaderConnection(currentConnection);
+    }
+
+    if (isReaderConnection()) {
+      return;
+    }
+
+    HostInfo readerHostInfo = getReaderHostInfo();
+    if (readerHostInfo == null) {
+      if (currentConnection == null || currentConnection.isClosed()) {
+        throw new SQLException(
+            Messages.getString("ReadWriteSplittingPlugin.5"),
+            MysqlErrorNumbers.SQL_STATE_UNABLE_TO_CONNECT_TO_DATASOURCE);
+      }
+
+      // reader connection host wasn't in our host list - stick with current connection
+      if (this.readerConnection != currentConnection) {
+        if (this.readerConnection != null) {
+          this.readerConnection.close();
+        }
+        this.readerConnection = currentConnection;
+      }
+      return;
+    }
+
+    syncSessionState(currentConnection, this.readerConnection, true);
+    this.currentConnectionProvider.setCurrentConnection(this.readerConnection, readerHostInfo);
+  }
+
+  private void initializeReaderConnection(JdbcConnection currentConnection) throws SQLException {
+    if (this.hosts.size() == 0) {
+      if (currentConnection == null || currentConnection.isClosed()) {
+        throw new SQLException(
+            Messages.getString("ReadWriteSplittingPlugin.4"),
+            MysqlErrorNumbers.SQL_STATE_UNABLE_TO_CONNECT_TO_DATASOURCE);
+      }
+
+      // stick with the current connection
+      if (this.readerConnection != currentConnection) {
+        if (this.readerConnection != null) {
+          this.readerConnection.close();
+        }
+        this.readerConnection = currentConnection;
+      }
```

Comment:
Can we create a method for the duplicated logic?

Reply:
Yep, good suggestion